pkg/config/validation/validation.go

```diff
@@ -2667,6 +2667,55 @@ func analyzeUnreachableHTTPRules(routes []*networking.HTTPRoute,
 			}
 		}
 	}
+
+	validateByPassedRouteForCatchAll(routes, reportUnreachable)
+}
+
+func validateByPassedRouteForCatchAll(routes []*networking.HTTPRoute, reportUnreachable func(ruleno string, reason string)) {
+	catchall := false
+	if len(routes) == 1 && (len(routes[0].Match) == 1 || len(routes[0].Match) == 0) {
+		return
+	}
+	for rulen, route := range routes {
+		if len(route.Match) == 0 && route.Mirror == nil && route.Fault == nil {
+			catchall = true
+		} else {
+			for i, match := range route.Match {
+				if isCatchAllMatch(match) && i != len(route.Match)-1 {
```

Comment:
Isn't that all `route.Match` are catch all, then the route is catch all?

Reply:
I'm afraid not, if there is one match of the route is catch all match, then it's catch all route and the match below the catchall matched will be bypassed. FYI https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L376-L388

I also updated some of existing testing specs like below one to make it backward compatible.

From
```
  - name: "send product to sample.foo"
    match:
    - uri:
        prefix: "/"
    - uri:
        exact: "/"  
      method:
        exact: GET
```
To
```
  - name: "send product to sample.foo"
    match:
    - uri:
        exact: "/"
    - uri:
        prefix: "/"
      method:
        exact: GET
```

Reply:
you are right

Reply:
The two YAML configs you provided above function 100% identically. We should not warn in these cases.

There are two lists: HTTPRoutes, and HTTPMatchRequests.

Above we have 1 HTTPRoute with 2 HTTPMatchRequest.

Within an HTTPRoute, the order of HTTPMatchRequest is irrelevant. If any match, the request route matches; it doesn't matter _which one_ matches.

**Across** HTTPRoutes is where it matters

Reply:
I refacored the test case to validate **across** HTTPRoutes.

But even within an HTTPRoute, if the catchall match presents in the first one, the remaining matches are skipped as well although from data plane perspective it still works.

Reply:
Thanks @howardjohn. Any further comment? If no could you help merge?

Reply:
ping @howardjohn